Exceptions.md

```diff
@@ -0,0 +1,199 @@
+It would be good to get rid of these two extra layers of wrappers.
+
+### Don't swallow exceptions if you can't handle them
+
+If an exception is thrown in the primitives layer, it should be possible to let it bubble up through the layers all the way up to the client code, because in many situations, only the client code knows how to deal with the exception appropriately.
+For instance, if the exception is a network failure, and we're in an application which has an offline mode, we don't want the lower layers to swallow the exception and return some dummy value, or terminate the program, but the exception should be passed on to the client app, which could just say, "ok so let's switch to offline mode".
+
+### Printing stack traces
+
+While a client is debugging their app, it would be useful for them to get a stack trace in case of unhandled exceptions which goes all the way through the stack, and is not interrupted by the Dafny layer.
+
+
+## Implementation
+
+The type `Result<T>` is hardcoded into Dafny, the same way as `seq`, `map`, `int`, etc.
+
+Given a `MethodCall(args)` which returns `Result<T>`, let's define
+
+```
+var v :- MethodCall(args);
+... rest of the code using v as if it had type T ...
+```
+
+to be syntactic sugar for
+
+```
+var res_v: Result<T> := MethodCall(args);
+if res_v.Failure? {
+	return Failure(res_v.error);
+}
+var v := res_v.value;
+... rest of the code using v as if it had type T ...
+```
+
+as far as verification and Boogie is concerned.
+
+We will not really implement it as syntactic sugar, though, because we want a separate AST node for this so that the compiler can regognize it easily.
+
+The compiler will then compile all methods which return `Result<T>` into C# methods with signatures of the form
+
+```
+public T MethodCall(A1 a1, A2 a2, ...) throws Exception
+```
+
+(Note that C# method signatures do not have `throws` clauses, but we just add them here for clarity, even though that's not valid C# code).
+
+In particular, the same signature translation also applies for methods annotated with `{:extern}`, so for instance, given the C# method `System.IO.Stream.Read` with signature 
+
+```
+public int Read (byte[] buffer, int offset, int count)
+```
+
+one could write a Dafny stub for it as
+
+```
+method {:extern} Read(buffer: array<byte>, offset: int32, count: int32) returns (nRead: Result<int32>)
+```
+
+which would enable us to call exception-throwing C# methods from Dafny without an extra unverified layer turning exceptions into `Result<int32>`.
+
+In the same way, methods implemented in Dafny which return `Result<T>` would be compiled to exception-throwing methods and thus be directly usable by C# client programs.
+
+Moreover,
+
+```
+return Failure(someErrorMsg);
+```
+
+is compiled into a `throw` statement:
+
+```
+throw new Exception(dafnyStringToNativeString(someErrorMsg));
```

Comment:
If instead of hardcoding the type `Result<T>`, we allow a user-defined type, then that user-defined datatype could have several constructors for the failure case. If we additionally allow some kind of `Throw` member of the datatype, then that `Throw` member could decide on different C# exceptions to throw, according to the constructor used.